Bumps dart SDK to 2.12 and FFI to stable version

pak3nuh opened the pull request:
The long awaited PR that enables null safety and stabilizes FFI.

This PR also bumps dependencies to the latest version that supports the SDK 2.12.

The bump isn't to a newer SDK to increase compatibility as none of the newer features are needed for now.

Closes #12

pak3nuh committed:
Update SDK WIP

pak3nuh committed:
Unit tests fixed

Updated dependencies to be compatible with dart 2.12
Fixed unit tests.

pak3nuh committed:
Fixes linter package

pak3nuh committed:
Cleanup integration tests

pak3nuh committed:
Bumps pact native library to 0.1.1

pak3nuh committed:
Updates CHANGELOG.md

ccosnean commented on lib/src/contract_builder_api.dart:
```diff
@@ -197,11 +193,11 @@ class RequestBuilder {
   Map<String, String> query = {};
 
-  ResponseBuilder get response => _response;
+  ResponseBuilder get response => _response ;
```
```suggestion
  ResponseBuilder get response => _response;
```

pak3nuh replied:
Will fix

ccosnean commented on pubspec.yaml:
```diff
@@ -1,18 +1,17 @@
 description: A consumer library to build and test pact contracts.
-version: 0.1.0
-# homepage: https://www.example.com
+version: 0.2.0
```
This is a breaking change, consider bumping it to `1.0.0`.
```suggestion
version: 1.0.0
```

pak3nuh replied:
Semver states modules may break API compability up until version 1.0.0 without bumping majors.
There is a section on the README that states the API might change up until stable version 1.

ccosnean commented on lib/src/pact_contract_dto.dart:
Would be neater to split each model into its own class [Like here](https://github.com/ccosnean/dart_pact_consumer/commit/f07164f6aa9f5b8a7dabf9afb455a8384889c1d9#diff-caaefeee6dbc6e375395275c4a4c9d069946d3afc706f3ca149c09e8c4e568f1). `pact_contract_dto` is not very explanatory...

pak3nuh replied:
This is a matter of personal preference.
Right now the json model is quite short. Maybe once it gets bigger I'll split it.

ccosnean commented on lib/src/pact_contract_dto.dart:
```diff
@@ -66,15 +67,15 @@ class ProviderState {
 @JsonSerializable()
-class Response {
-  int status;
+class Response implements ReifiedJson {
```
Response is missing `matchingRules` [here](https://github.com/ccosnean/dart_pact_consumer/commit/f07164f6aa9f5b8a7dabf9afb455a8384889c1d9#diff-fb08a617314635784c6e4036ef4b764c6717e8c9a44aa6f442bdde0fd223735c) is how i did it in the past.

it is really important when you try to match the `Location` header after a `POST` request, for example. I'm sure there are other use cases...

pak3nuh replied:
This PR focus on update the dart SDK to be used on null safe environments.
I'll ask you to create an issue for a feature request to add it in a future version please.

ccosnean commented on lib/src/pact_contract_dto.dart:
```diff
@@ -83,16 +84,16 @@ class Response {
 
-  String method;
-  String path;
+  late String method;
+  late String path;
```
The Request `method` cannot be null. ever.

Request `path`, `body`, `query` and `headers` can be...

i know you try to accommodate for dart's `..` operator but this is too risky. You will run into runtime errors, that is really bad. (not really null-safe)

better to catch all errors at compile time. [Like this](https://github.com/ccosnean/dart_pact_consumer/commit/f07164f6aa9f5b8a7dabf9afb455a8384889c1d9#diff-05239619bdb97dc53d5e6a64a5c0043e28b75d47be036cf62c988b0bb05d5a46)

<img width="1040" alt="Screenshot 2023-03-17 at 10 47 57 PM" src="https://user-images.githubusercontent.com/22283233/226042124-bcf42a86-446c-4b2c-8181-1015435261c8.png">

pak3nuh replied:
I agree with you that we should uplift error to compile time.
The thing is that classes in that file is a DTO to transfer Json structures. It is not inteded to be used directly. In fact I think it's not exported to the public API.
I use late here because techincally every object can be null comming from json. I would need to have a null assertion when getting every value and it seems like clutter.
Does it make sense?

ccosnean commented on integration_test/pact_host_client_test.dart:
```diff
@@ -4,7 +4,7 @@ import 'package:test/test.dart';
 void main() {
   group('PactHost', () {
-    const host = 'http://localhost:9292';
+    const host = 'http://192.168.96.1:9292';
```
`192.168.96.1` is a private IP address and directly related to your Wi-Fi network, are you sure this is right?

pak3nuh replied:
Nice catch, it shouldn't be here.
This is an ip to a local pact broker used in testing.
Do you know any form of passing args to the test runner? I've checked the dart help manual and I haven't found any.

ccosnean replied:
Yes, Using the `--dart-define` flag, [Like this](https://dartcode.org/docs/using-dart-define-in-flutter/)

```bash
flutter test --dart-define PACT_HOST='192.168.96.1'
```

and later in code you can get it with

```dart
const host = String.fromEnvironment('PACT_HOST');
```

pak3nuh replied:
Yeah that is a good suggestion, will use.

BTW, dart executable doesn't have the `dart-define` switch. I guess that if flutter trickery that sets up variables in the process it spawns. At least that I can use in the test command

ccosnean replied:
In pure dart it is just `--define` [described here](https://dartcode.org/docs/using-define-in-dart/)

pak3nuh replied:
This is configuration targeting vscode's `launch.json`. The dart SDK has no knowledge of such argument.

But I'll stick to the environment variable. It's a nice solution regardeless.

pak3nuh committed:
Applies review comments

Removes local addresses from test code. Runners should
define a PACT_HOST env variable with the broker address.
Implements close method on the PactHost client. It was
holding connections open and tests didn't finish.
